wallet-new/src/Cardano/Wallet/Kernel/Restore.hs

```diff
@@ -285,13 +313,25 @@ restoreWalletHistoryAsync wallet rootId target tgtSlot prefilter = do
     withNode :: forall a. (NodeConstraints => WithNodeState IO a) -> IO a
     withNode action = withNodeState (wallet ^. walletNode) (\_lock -> action)
 
--- Update the restoration information for a wallet.
+-- Update the restoration information for a wallet. If somebody else is holding
+-- the 'walletRestorationTask' MVar, we'll just skip this; the worst that can
+-- happen is the restoration's progress information is not quite up-to-date.
+--
+-- The idea is that the wallet restoration should be thought of
+-- as properly nested inside the WalletRestorationInfo's MVar. So to avoid
+-- deadlock, we can't grab the WalletRestorationInfo MVar unconditionally here.
+--
```

Comment:
I think this comment is out-of-date and I'll edit, but I *do* think it makes sense to avoid `modifyMVar_` in favor of `tryTakeMVar` here. I'd rather see a progress bar that sometimes failed to update promptly than a potential deadlock.